File: src/statesman/commands/ConditionalCommand.java
package statesman.commands;

import java.util.ArrayList;

import statesman.GameException;

public class ConditionalCommand extends Command {

    public static final String ID_CONDITIONAL = "cond";
    public static final String DELIMITER_OR = "||";
    public static final String DELIMITER_AND = "&&";

    protected CommandGroup _group;
    protected CommandGroup _elseGroup;

    private ArrayList<String> _targetNames;
    private ArrayList<Boolean> _targetValues;

    private boolean _orMode;
    private Boolean _shouldExecute;

    private ConditionalCommand() {
        _group = null;
        _elseGroup = null;
        _orMode = false;
        _targetNames = null;
        _targetValues = null;
        _shouldExecute = null;
    }

    public ConditionalCommand(CommandGroup group, CommandGroup elseGroup,
            ArrayList<String> targetNames, ArrayList<Boolean> targetValues, boolean orMode) {
        this();
        _group = group;
        _elseGroup = elseGroup;
        _targetNames = targetNames;
        _targetValues = targetValues;
        _orMode = orMode;
    }

    private static Command _defaultInstance;

    public static Command getDefault() {
        if (_defaultInstance == null) {
            _defaultInstance = new ConditionalCommand();
        }
        return _defaultInstance;
    }

    @Override
    public Command fromText(String commandId, String[] arguments) throws GameException {
        if (arguments.length >= 2) {
            Boolean orMode = null;
            ArrayList<Boolean> targetValues = new ArrayList<Boolean>();
            ArrayList<String> targetNames = new ArrayList<String>();

            // The first argument is the command name, so start evaluating
            // the parts of the conditional with the second argument.
            boolean isConditionNext = false;
            boolean isConditionPrevious = false;
            for (int i = 1; i < arguments.length; i++) {
                String operatorOrTargetName = arguments[i].trim();

                boolean foundOrOperator =
                        operatorOrTargetName.equalsIgnoreCase(DELIMITER_OR);
                boolean foundAndOperator =
                        operatorOrTargetName.equalsIgnoreCase(DELIMITER_AND);

                if (operatorOrTargetName.length() == 2 && isConditionNext)
                {
                    if (!foundOrOperator && !foundAndOperator)
                    {
                        throw new GameException("Unknown operator found.");
                    }

                    if (orMode == null)
                    {
                        orMode = foundOrOperator;
                    }
                    else if ((orMode.booleanValue() && foundAndOperator)
                            || (!orMode.booleanValue() && foundOrOperator))
                    {
                        throw new GameException(
                                "Combining and/or conditional operators are not allowed.");
                    }
                    isConditionNext = false;
                    isConditionPrevious = true;
                    continue;
                }

                if (foundOrOperator || foundAndOperator)
                {
                    throw new GameException("Incorrect condition order.");
                }

                // Check for the negation operator.
                boolean targetValue = !operatorOrTargetName.startsWith("!");
                if (!targetValue)
                {
                    operatorOrTargetName = operatorOrTargetName.substring(1);
                }
                targetNames.add(operatorOrTargetName);
                targetValues.add(targetValue);
                isConditionNext = true;
                isConditionPrevious = false;
            }

            if (isConditionPrevious)
            {
                throw new GameException("Found stray operator inside condition.");
            }

            // Assume Or mode if it's a single condition.
            if (orMode == null)
            {
                orMode = true;
            }

            return new ConditionalCommand(
                    new CommandGroup(""),
                    new CommandGroup(""),
                    targetNames,
                    targetValues,
                    orMode);
        }
        return null;
    }

    @Override
    public void execute() {
        for (int i = 0; i < _targetNames.size(); i++) {
            String targetName = _targetNames.get(i);
            boolean targetValue = _targetValues.get(i);
            boolean keyValue = JumpCommand.getConditionValue(targetName);

            boolean stopLooping = updateState(keyValue == targetValue);
            if (stopLooping) {
                break;
            }
        }

        if (_shouldExecute) {
            _group.execute();
        } else {
            _elseGroup.execute();
        }
        _shouldExecute = null;
    }

    private boolean updateState(boolean newState) {
        if (newState) {
            if (_orMode) {
                _shouldExecute = true;
                return true;
            }
            if (_shouldExecute == null) {
                _shouldExecute = true;
            }
        } else {
            _shouldExecute = false;
        }
        return false;
    }

    public CommandGroup getGroup() {
        return _group;
    }

    public CommandGroup getElseGroup() {
        return _elseGroup;
    }

    public ArrayList<String> getTargetNames() {
        return _targetNames;
    }

    public ArrayList<Boolean> getTargetValues() {
        return _targetValues;
    }

}

File: src/Commands/ConditionalCommand.cs
namespace Statesman.Commands
{
    public class ConditionalCommand : Command
    {
        public const string CommandConditional = "cond";
        private const string OrDelimiter = "||";
        private const string AndDelimiter = "&&";

        public List<string> TargetNames { get; }
        public List<bool> TargetValues { get; private set; }
        public Function Group { get; private set; }
        public Function ElseGroup { get; private set; }

        private readonly bool _orMode;
        private bool? _executeTrueGroup;

        public ConditionalCommand(
            Function group,
            Function elseGroup,
            List<string> targetNames,
            List<bool> targetValues,
            bool orMode)
        {
            Group = group;
            ElseGroup = elseGroup;
            TargetNames = targetNames;
            TargetValues = targetValues;
            _executeTrueGroup = null;
            _orMode = orMode;
        }

        public new static Command FromText(string commandName, string[] arguments)
        {
            if (commandName != CommandConditional)
            {
                return null;
            }
            if (arguments.Length >= 2)
            {
                bool? orMode = null;
                List<string> targetNames = new();
                List<bool> targetValues = new();

                // The first argument is the command name, so start evaluating
                // the parts of the conditional with the second argument.
                bool isConditionNext = false;
                bool isConditionPrevious = false;
                for (int i = 1; i < arguments.Length; i++)
                {
                    string operatorOrTargetName = arguments[i].Trim();

                    bool foundOrOperator = operatorOrTargetName.Equals(OrDelimiter);
                    bool foundAndOperator = operatorOrTargetName.Equals(AndDelimiter);

                    if (operatorOrTargetName.Length == 2 && isConditionNext)
                    {
                        if (!foundOrOperator && !foundAndOperator)
                        {
                            throw new GameException("Unknown operator found.");
                        }

                        if (!orMode.HasValue)
                        {
                            orMode = foundOrOperator;
                        }
                        else if ((orMode.Value && foundAndOperator) || (!orMode.Value && foundOrOperator))
                        {
                            throw new GameException("Combining and/or conditional operators are not allowed.");
                        }
                        isConditionNext = false;
                        isConditionPrevious = true;
                        continue;
                    }

                    if (foundOrOperator || foundAndOperator)
                    {
                        throw new GameException("Incorrect condition order.");
                    }

                    // Check for the negation operator.
                    bool targetValue = !operatorOrTargetName.StartsWith("!");
                    if (!targetValue)
                    {
                        operatorOrTargetName = operatorOrTargetName.Substring(1);
                    }
                    targetNames.Add(operatorOrTargetName);
                    targetValues.Add(targetValue);
                    isConditionNext = true;
                    isConditionPrevious = false;
                }

                if (isConditionPrevious)
                {
                    throw new GameException("Found stray operator inside condition.");
                }

                // Assume Or mode if it's a single condition.
                if (!orMode.HasValue)
                {
                    orMode = true;
                }

                return new ConditionalCommand(
                        new Function(""),
                        new Function(""),
                        targetNames,
                        targetValues,
                        orMode.Value);
            }
            return null;
        }

        public override void Execute()
        {
            // Evaluate all conditions.
            for (int i = 0; i < TargetNames.Count; i++)
            {
                bool currentValue = JumpCommand.GetConditionValue(TargetNames[i]);
                bool condition = currentValue == TargetValues[i];
                if (UpdateState(condition))
                {
                    break;
                }
            }
            // Execute the commands inside the condition's scope depending
            // on whether the condition is true or false.
            if (_executeTrueGroup.Value)
            {
                Group.Execute();
            }
            else
            {
                ElseGroup.Execute();
            }
            // Reset the state.
            _executeTrueGroup = null;
        }

        private bool UpdateState(bool condition)
        {
            // The condition is true.
            if (condition)
            {
                // Or mode: requires only one condition to be true.
                if (_orMode)
                {
                    _executeTrueGroup = true;
                    return true;
                }
                // And mode: the first condition is true.
                else if (_executeTrueGroup == null)
                {
                    _executeTrueGroup = true;
                }
            }
            else
            {
                // The condition is false.
                _executeTrueGroup = false;
                // There's no point in checking the other conditions if we're
                // not in or mode.
                return !_orMode;
            }
            return false;
        }
    }
}